internal/objectstore/fileobjectstore.go

```diff
@@ -387,63 +418,75 @@ func (t *fileObjectStore) loop() error {
+	t.logger.Debugf(ctx, "getting object %q from file storage", path)
 
 	metadata, err := t.metadataService.GetMetadata(ctx, path)
 	if errors.Is(err, domainobjectstoreerrors.ErrNotFound) {
 		return nil, -1, errors.Errorf("get metadata: %w", objectstoreerrors.ObjectNotFound)
-	}
-	if err != nil {
+	} else if err != nil {
 		return nil, -1, errors.Errorf("get metadata: %w", err)
 	}
 
-	return t.getWithMetadata(ctx, metadata)
+	return t.getWithMetadata(ctx, metadata, fallbackStrategy)
 }
 
 func (t *fileObjectStore) getBySHA256(ctx context.Context, sha256 string) (io.ReadCloser, int64, error) {
-	t.logger.Debugf(context.TODO(), "getting object with SHA256 %q from file storage", sha256)
+	t.logger.Debugf(ctx, "getting object with SHA256 %q from file storage", sha256)
 
 	metadata, err := t.metadataService.GetMetadataBySHA256(ctx, sha256)
 	if errors.Is(err, domainobjectstoreerrors.ErrNotFound) {
 		return nil, -1, errors.Errorf("get metadata by SHA256: %w", objectstoreerrors.ObjectNotFound)
-	}
-	if err != nil {
+	} else if err != nil {
 		return nil, -1, errors.Errorf("get metadata by SHA256: %w", err)
 	}
 
-	return t.getWithMetadata(ctx, metadata)
+	// Getting a file by SHA256 implies that we're looking for a file that
+	// has a specific hash. If we can't find the file, then we should return
+	// an error, as we can't fallback to a remote source. This is a hard
+	// back stop.
+	// If we do want to allow for a remote fallback, we should expose that
+	// configuration option all the way to the objectstore interface. For now,
+	// this is non-configurable.
+	// The added benefit is that this prevents infinite loops, where we keep
+	// trying to get the file from the remote source, but it doesn't exist.
+	return t.getWithMetadata(ctx, metadata, NoFallback)
 }
 
-func (t *fileObjectStore) getBySHA256Prefix(ctx context.Context, sha256 string) (io.ReadCloser, int64, error) {
-	t.logger.Debugf(context.TODO(), "getting object with SHA256 %q from file storage", sha256)
+func (t *fileObjectStore) getBySHA256Prefix(ctx context.Context, sha256 string, fallbackStrategy FallbackStrategy) (io.ReadCloser, int64, error) {
+	t.logger.Debugf(ctx, "getting object with SHA256 %q from file storage", sha256)
 
 	metadata, err := t.metadataService.GetMetadataBySHA256Prefix(ctx, sha256)
 	if errors.Is(err, domainobjectstoreerrors.ErrNotFound) {
 		return nil, -1, errors.Errorf("get metadata by SHA256 prefix: %w", objectstoreerrors.ObjectNotFound)
-	}
-	if err != nil {
+	} else if err != nil {
 		return nil, -1, errors.Errorf("get metadata by SHA256 prefix: %w", err)
 	}
 
-	return t.getWithMetadata(ctx, metadata)
+	return t.getWithMetadata(ctx, metadata, fallbackStrategy)
 }
 
-func (t *fileObjectStore) getWithMetadata(ctx context.Context, metadata objectstore.Metadata) (io.ReadCloser, int64, error) {
+func (t *fileObjectStore) getWithMetadata(ctx context.Context, metadata objectstore.Metadata, fallbackStrategy FallbackStrategy) (io.ReadCloser, int64, error) {
```

Comment:
How about giving a slice of(even now, either the remote one or nil, but it will be easily extensible for the future) the real-getter functions instead of a const.
```
func (t *fileObjectStore) getWithMetadata(
    ctx context.Context, metadata objectstore.Metadata, 
    fallbackStrategies ...Retriever) (io.ReadCloser, int64, error) {
```

Reply:
I'll consider that in my next PR, which gets the blobs when the metadata changes. Thanks for the feedback.